### PR TITLE
Try again: in unix like system, set foreground process while running external command

### PR DESCRIPTION
**Pull request by WindSoilder**
# Description

Relative: #6156 #3861 #6240

Check the implementation in [ion](https://gitlab.redox-os.org/redox-os/ion/-/tree/master/)

I've done running unit testing on unix several times, and done some manual tests, and it works for me.

@herlon214 Would you please check if the build fix the issue #6240 ?

# Tests

Make sure you've done the following:

- [ ] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [X] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [X] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [X] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

**Comment by herlon214** [comment]
@WindSoilder looks like your branch is working fine here 🎉

**Comment by fdncred** [comment]
I tested this too on my mac and couldn't get it to suspend like it was doing.

**Comment by fdncred** [comment]
@jntrnr I requested your review here because you had some way of reproducing some failure with testing. Can you please repeat that on this PR to see if this PR fixes the issues you were seeing?

**Comment by jntrnr** [comment]
Testing it out on my machine, it looks like this doesn't crash anymore which is great.

I'm trying to get the functionality to work in GNOME Terminal: if I type `vim foo.txt` I'd expect to see that as part of the title, right? This is how it works in fish, but with your changes the title doesn't seem to change. Am I correct that this should be changing?

This change:
Screenshot from 2022-08-12 06-32-40

Fish:
Screenshot from 2022-08-12 06-33-13

**Comment by fdncred** [comment]
@jntrnr I believe updating the title is a separate thing since right now that is done via ansi escape sequences. I thought this was more for the tmux usage. This issue https://github.com/nushell/nushell/issues/6156 goes into detail about what the use case is because I kept pushing back on saying that we can just set the title easily before I understood what the goal was. However, we could throw the escape code in there when this happens too, if we wanted to.

**Comment by WindSoilder** [comment]
> I thought this was more for the tmux usage

Yeah, and I've checked `ion`, it doesn't set title as well

**Comment by fdncred** [comment]
@rosshadden can you please test this to make sure it works as described in #6156?

**Comment by rosshadden** [comment]
Yeah I'll try when I can. I just cut my thumb and got stitches tonight so it might have to wait a day or so 😅

**Comment by rosshadden** [comment]
A bit late since you merged it a couple hours ago, but I can confirm this works too. Props for getting it working again while fixing the issue that was found with the other approach!

**Comment by fdncred** [comment]
thanks @rosshadden. we appreciate your feedback and testing!

**Comment by dandavison** [comment]
I've come across a bug introduced by this commit (1d18f6947ed092a5adb36c327f7eccb4436130a3). The only way I currently know to reproduce it involves less, e.g.

```
yes | head -n 999 | with-env [LESS ""] { less }
``` 

Assuming you don't have a very large screen, that command should leave the less process running. It should exit with SIGINT, and it does on the previous commit (df3b6d9d264747bff0e2a81ab6eddb063d11a442), but does not on 1d18f6947ed092a5adb36c327f7eccb4436130a3.

I'm using `less 600` on MacOS.

**Comment by fdncred** [comment]
@dandavison How can you tell if it's exiting with SIGINT? or are you just saying it doesn't respond to Ctrl-C?

**Comment by dandavison** [comment]
I'm sorry, my initial report was confused: the bug is actually that it doesn't respond to the `q` keyboard shortcut of `less`, which should cause the less process to exit (and is what users will usually use).

----------------------------------------

For signals, the behavior is the same in bash and zsh and nushell (1d18f6947ed092a5adb36c327f7eccb4436130a3) and so there is no bug relating to the way it handles signals. So the following is all irrelevant but, for the record, on bash, zsh, and nushell, it does not respond to `SIGINT` (e.g. `/bin/kill -s INT` or `kill -s 2` using nushell's built-in),  but it does exit with `KILL` (e.g. `/bin/kill -s KILL` or `kill -s 9` in nushell).

**Comment by fdncred** [comment]
hmmm, i don't have a problem with less 590 responding to q with your example.

**Comment by dandavison** [comment]
@fdncred, hm, I certainly hope I'm not wasting time here! But I do still think there's a bug somewhere. I just tested on two separate MacOS machines. With a nu binary built on current main (f1d72e2670d759ef7556d680836113334bd88145), the bug reproduced on both machines. I tested less 590 and less 600 on one of those machines and it reproduced for both less versions.

**Comment by unrelentingtech** [comment]
I think the missing piece might be not unblocking signals in the forked child (`pre_exec`). [ion does it here](https://gitlab.redox-os.org/redox-os/ion/-/blob/814ed5a887ad88b98b142af28f653ff4f249de6c/src/lib/shell/pipe_exec/fork.rs#L27-31)

**Comment by dandavison** [comment]
@unrelentingtech thanks for looking into this. I haven't experimented with what you said yet. However, the bug seemed to be present in ion also (@WindSoilder and I both observed that I believe).

**Comment by unrelentingtech** [comment]
Looking at fish, [here](https://github.com/fish-shell/fish-shell/blob/14d2a6d8ff18434a9672887d60009eef92ea21cc/src/postfork.cpp#L164-L182) it actually resets signals for the forked child before launch. It's not super obvious at a first glance but it's a `sigprocmask(SIG_SETMASK)` with a set that's normally empty [or has INT+QUIT when job control is off or something](https://github.com/fish-shell/fish-shell/blob/50d37527a9f8377979bcc0605c7109a3f59af257/src/exec.cpp#L388-L394). Note that `SIG_SETMASK` is key here, that's why the previously blocked signals go away.

File: crates/nu-system/src/foreground.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                // https://github.com/fish-shell/fish-shell/blob/023042098396aa450d2c4ea1eb9341312de23126/src/exec.cpp#L398
+                if let Err(e) = unistd::setpgid(Pid::from_raw(0), Pid::from_raw(0)) {
+                    println!("ERROR: setpgid for external failed, result: {e:?}");
+                }
+                Ok(())
+            });
+        }
+    }
+
+    // If `prepare_to_foreground` function is not called, the function will fail with silence and do nothing.
```
Comment (unrelentingtech) [inline crates/nu-system/src/foreground.rs:83]: I'm not sure how that comment is true btw? The check is for stdin being a tty; `prepare_to_foreground` does not touch stdin at all…

File: crates/nu-system/src/foreground.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            if let Err(e) = nix::unistd::tcsetpgrp(nix::libc::STDIN_FILENO, unistd::getpgrp()) {
+                println!("ERROR: reset foreground id failed, tcsetpgrp result: {e:?}");
+            }
+        }
+        unblock()
+    }
+}
+
+// TODO: investigate if we can set foreground process through windows system call.
+#[cfg(target_family = "windows")]
```
Comment (unrelentingtech) [inline crates/nu-system/src/foreground.rs:109]: The blank impl probably should be `cfg(not(unix))` rather than windows specifically…

**Comment by unrelentingtech** [comment]
Turns out there's some signal manipulation happening in `std::process::Command`, it probably was unblocking them. Either way I'm trying to add more stuff from the GNU manual now. Getting interesting but not correct results so far.

**Pull request merged**